modules/light-clients/10-wasm/types/client_state.go

```diff
@@ -0,0 +1,662 @@
+	}
+	output := clientStateCallResponse{}
+	if err := json.Unmarshal(out.Data, &output); err != nil {
+		return nil, sdkerrors.Wrapf(ErrUnableToUnmarshalPayload, fmt.Sprintf("underlying error: %s", err.Error()))
+	}
+	if !output.Result.IsValid {
+		return nil, fmt.Errorf("%s error occurred while updating client state", output.Result.ErrorMsg)
+	}
+
+	output.resetImmutables(c)
+	return output.NewClientState, nil
+}
+
+func (c *ClientState) VerifyUpgradeAndUpdateState(ctx sdk.Context, cdc codec.BinaryCodec, store sdk.KVStore, newClient exported.ClientState, newConsState exported.ConsensusState, proofUpgradeClient, proofUpgradeConsState []byte) (exported.ClientState, exported.ConsensusState, error) {
+	wasmUpgradeConsState, ok := newConsState.(*ConsensusState)
+	if !ok {
+		return nil, nil, sdkerrors.Wrapf(clienttypes.ErrInvalidConsensus, "upgraded consensus state must be Tendermint consensus state. expected %T, got: %T",
+			&ConsensusState{}, wasmUpgradeConsState)
+	}
+
+	// last height of current counterparty chain must be client's latest height
+	lastHeight := c.LatestHeight
+	lastHeightConsensusState, err := GetConsensusState(store, cdc, lastHeight)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "could not retrieve consensus state for lastHeight")
+	}
+
+	const VerifyUpgradeAndUpdateState = "verifyupgradeandupdatestate"
+	payload := make(map[string]map[string]interface{})
+	payload[VerifyUpgradeAndUpdateState] = make(map[string]interface{})
+	inner := payload[VerifyUpgradeAndUpdateState]
+	inner["me"] = c
+	inner["new_client_state"] = newClient
+	inner["new_consensus_state"] = newConsState
+	inner["client_upgrade_proof"] = proofUpgradeClient
+	inner["consensus_state_upgrade_proof"] = proofUpgradeConsState
+	inner["last_height_consensus_state"] = lastHeightConsensusState
+
+	encodedData, err := json.Marshal(payload)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrapf(ErrUnableToMarshalPayload, fmt.Sprintf("underlying error: %s", err.Error()))
+	}
+	out, err := callContract(c.CodeId, ctx, store, encodedData)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrapf(ErrUnableToCall, fmt.Sprintf("underlying error: %s", err.Error()))
+	}
+	output := clientStateCallResponse{}
+	if err := json.Unmarshal(out.Data, &output); err != nil {
+		return nil, nil, sdkerrors.Wrapf(ErrUnableToUnmarshalPayload, fmt.Sprintf("underlying error: %s", err.Error()))
+	}
+	if !output.Result.IsValid {
+		return nil, nil, fmt.Errorf("%s error occurred while updating client state", output.Result.ErrorMsg)
+	}
+	output.resetImmutables(c)
+	return output.NewClientState, output.NewConsensusState, nil
+}
+
+func (c *ClientState) ZeroCustomFields() exported.ClientState {
+	const ZeroCustomFields = "zerocustomfields"
+	payload := make(map[string]map[string]interface{})
+	payload[ZeroCustomFields] = make(map[string]interface{})
+	inner := payload[ZeroCustomFields]
+	inner["me"] = c
+
+	encodedData, err := json.Marshal(payload)
+	if err != nil {
+		// TODO: Handle error
```

Comment:
There are a few places in this file where I left the error handling marked with `TODO`. Some methods signature doesn't allow to return an error, so options are:
* `panic` on error (not great)
* modify the interface / method signature (broader change)

It would be great if you could advice here, thanks.